flink-table/flink-table-common/src/main/java/org/apache/flink/table/sources/PartitionableTableSource.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.sources;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Adds support for partition pruning to a [[TableSource]].
+ * A {@link TableSource} extending this interface is a partition table and able to
+ * prune partitions before reading data.
+ *
+ * <p>Partitions is represented as a {@code Map<String, String>} which maps from partition-key
+ * to value. The partition columns and values are NOT of strict order in the Map. The correct
+ * order of partition fields should be provided by {@link #getPartitionKeys()}.
+ */
+public interface PartitionableTableSource {
+
+	/**
+	 * Gets all partitions as key-value map belong to this table.
+	 */
+	List<Map<String, String>> getAllPartitions();
+
+	/**
+	 * Get the partition keys of the table. This should be an empty set if the table
+	 * is not partitioned.
+	 *
+	 * @return partition keys of the table
+	 */
+	List<String> getPartitionKeys();
+
+	/**
+	 * Return the flag to indicate whether partition pruning has been tried. Must return true on
+	 * the returned instance of {@link #applyPartitionPruning(List)}.
+	 */
+	boolean isPartitionPruned();
+
+	/**
+	 * Applies the remaining partitions to the table source. The {@code remainingPartitions} is
+	 * the remaining partitions of {@link #getAllPartitions()} after partition pruning applied.
+	 *
+	 * <p>After trying to apply partition pruning, we should return a new {@link TableSource}
+	 * instance which holds all pruned-partitions. Even if we actually pruned nothing,
+	 * it is recommended that we still return a new {@link TableSource} instance since we will
+	 * mark the returned instance as partition pruned and will never try again.
+	 *
+	 * @param remainingPartitions Remaining partitions after partition pruning applied.
+	 * @return A new cloned instance of {@link TableSource}.
+	 */
+	TableSource applyPartitionPruning(List<Map<String, String>> remainingPartitions);
+
+	/**
+	 * Whether drop partition predicates after apply partition pruning.
```

Comment:
could you explain it more clearly why we need this method ?

Reply:
if the values of partition fields in a partition are not unique, the partition predicates could not be dropped after partition pruning, otherwise the result is incorrect. 
e.g. a table source is partition by hour,  and its partition field is `timestamp` which precision is milliseconds. each partition contains multiple `timestamp` values. we also could use partition predicates to filter partitions, but the remaining partitions may contains unexpected result which could not meet the partition predicates. the partition predicates could be reserved.
however, how to filter partitions is very special too. 

I think we should delete this method and require that the values of partition fields in every partition should be unique. otherwise use should implement another partition prune rule to meet special needs.

Reply:
I agree. `supportDropPartitionPredicate()` has been removed.